projects/kit/components/progress/progress-bar/progress-color-segments.directive.ts

```diff
@@ -1,45 +1,53 @@
-import {Directive, inject, Input} from '@angular/core';
+import {computed, Directive, inject, Input, signal} from '@angular/core';
 import {toSignal} from '@angular/core/rxjs-interop';
-import {ResizeObserverService} from '@ng-web-apis/resize-observer';
-import {tuiWatch, tuiZonefull} from '@taiga-ui/cdk/observables';
+import {
+    MutationObserverService,
+    WA_MUTATION_OBSERVER_INIT,
+} from '@ng-web-apis/mutation-observer';
 import {tuiInjectElement} from '@taiga-ui/cdk/utils/dom';
-import {BehaviorSubject, combineLatest, distinctUntilChanged, map} from 'rxjs';
+import {map} from 'rxjs';
 
 @Directive({
     standalone: true,
     selector: 'progress[tuiProgressBar][tuiProgressColorSegments]',
-    providers: [ResizeObserverService],
+    providers: [
+        MutationObserverService,
+        {
+            provide: WA_MUTATION_OBSERVER_INIT,
+            useValue: {
+                attributeOldValue: true,
+            },
+        },
+    ],
     host: {'[style.--tui-progress-color]': 'color()'},
 })
 export class TuiProgressColorSegments {
-    private readonly colors$ = new BehaviorSubject<string[]>([]);
+    private readonly colors = signal<readonly string[]>([]);
     private readonly el = tuiInjectElement<HTMLProgressElement>();
+    private readonly position = toSignal(
+        inject(MutationObserverService, {self: true}).pipe(map(() => this.el.position)),
+        {initialValue: this.el.position},
+    );
 
-    protected readonly color = toSignal(
-        combineLatest([
-            this.colors$,
-            inject(ResizeObserverService, {self: true}).pipe(
-                map(() => this.el.offsetWidth),
-                distinctUntilChanged(),
-            ),
-        ]).pipe(
-            map(([colors, width]) => {
-                const segmentWidth = Math.ceil(width / colors.length);
-                const colorsString = colors.reduce(
-                    (acc, color, i) =>
-                        `${acc}, ${color} ${i * segmentWidth}px ${(i + 1) * segmentWidth}px`,
-                    '',
-                );
+    protected readonly color = computed(() => {
+        const colors = this.colors();
+        const position = this.position();
 
-                return `linear-gradient(to right ${colorsString})`;
-            }),
-            tuiZonefull(),
-            tuiWatch(),
-        ),
-    );
+        if (!colors.length || position <= 0) {
+            return null;
+        }
+
+        const colorsString = colors.reduce(
+            (acc, color, i) =>
+                `${acc}, ${color} calc(${i} / ${colors.length} * 100% / ${position}) calc(${i + 1} / ${colors.length} * 100% / ${position})`,
+            '',
+        );
+
+        return `linear-gradient(to right${colorsString})`;
+    });
 
     @Input('tuiProgressColorSegments')
-    public set colors(colors: string[]) {
-        this.colors$.next(colors);
+    public set colorsSetter(colors: readonly string[]) {
+        this.colors.set(colors);
```

Comment:
Let's use modern signal-based input `import {input} from '@angular/core'` instead of setter+signal